File: LzssCompression.py
from dataclasses import dataclass
from typing import Optional
import array

@dataclass
class LzssSettings:
    frame_size: int = 0x1000
    frame_fill: int = 0
    frame_init_pos: int = 0xFEE
    max_match_length: int = 0x12
    min_match_length: int = 2

class LZSSEncoder:
    def __init__(self, settings: Optional[LzssSettings] = None):
        settings = settings if settings else LzssSettings()
        self.settings = settings
        self.N = settings.frame_size  # Size of ring buffer
        self.F = settings.max_match_length  # Upper limit for match_length
        self.THRESHOLD = settings.min_match_length  # Encode string into position and length if match_length is greater than this
        self.NIL = self.N  # Index for root of binary search trees
        self.PADDING = settings.frame_fill  # Padding character

        # Ring buffer
        self.text_buf = array.array('B', [self.PADDING] * (self.N + self.F - 1))

        # Tree structures
        self.lchild = array.array('i', [0] * (self.N + 1))
        self.rchild = array.array('i', [0] * (self.N + 257))
        self.parent = array.array('i', [0] * (self.N + 1))

        # Match properties
        self.match_position = 0
        self.match_length = 0

        self._init_state()

    def _init_state(self):
        """Initialize encoding state and trees"""
        for i in range(self.N + 1, self.N + 257):
            self.rchild[i] = self.NIL
        for i in range(self.N):
            self.parent[i] = self.NIL

    def _insert_node(self, r: int):
        """Insert string of length F into one of the trees"""
        cmp_val = 1
        key = self.text_buf[r:r + self.F]
        p = self.N + 1 + key[0]
        self.rchild[r] = self.lchild[r] = self.NIL
        self.match_length = 0

        while True:
            if cmp_val >= 0:
                if self.rchild[p] != self.NIL:
                    p = self.rchild[p]
                else:
                    self.rchild[p] = r
                    self.parent[r] = p
                    return
            else:
                if self.lchild[p] != self.NIL:
                    p = self.lchild[p]
                else:
                    self.lchild[p] = r
                    self.parent[r] = p
                    return

            i = 1
            while i < self.F:
                cmp_val = key[i] - self.text_buf[p + i]
                if cmp_val != 0:
                    break
                i += 1

            if i > self.match_length:
                self.match_position = p
                self.match_length = i
                if i >= self.F:
                    break

        self.parent[r] = self.parent[p]
        self.lchild[r] = self.lchild[p]
        self.rchild[r] = self.rchild[p]
        self.parent[self.lchild[p]] = r
        self.parent[self.rchild[p]] = r

        if self.rchild[self.parent[p]] == p:
            self.rchild[self.parent[p]] = r
        else:
            self.lchild[self.parent[p]] = r
        self.parent[p] = self.NIL

    def _delete_node(self, p: int):
        """Delete node p from tree"""
        if self.parent[p] == self.NIL:
            return

        if self.rchild[p] == self.NIL:
            q = self.lchild[p]
        elif self.lchild[p] == self.NIL:
            q = self.rchild[p]
        else:
            q = self.lchild[p]
            if self.rchild[q] != self.NIL:
                while self.rchild[q] != self.NIL:
                    q = self.rchild[q]
                self.rchild[self.parent[q]] = self.lchild[q]
                self.parent[self.lchild[q]] = self.parent[q]
                self.lchild[q] = self.lchild[p]
                self.parent[self.lchild[p]] = q
            self.rchild[q] = self.rchild[p]
            self.parent[self.rchild[p]] = q

        self.parent[q] = self.parent[p]
        if self.rchild[self.parent[p]] == p:
            self.rchild[self.parent[p]] = q
        else:
            self.lchild[self.parent[p]] = q
        self.parent[p] = self.NIL

def compress(data: bytes, settings: Optional[LzssSettings] = None) -> Optional[bytes]:
    """
    Compress the input data using LZSS algorithm.

    Args:
        data: Input bytes to compress
        settings: LzssSettings instance with compression parameters

    Returns:
        Compressed data or None if input data is empty
    """
    settings = settings if settings else LzssSettings()
    
    if not data:
        return None

    encoder = LZSSEncoder(settings)
    code_buf = bytearray(17)
    compressed = bytearray()

    code_buf[0] = 0
    code_buf_ptr = 1
    mask = 1

    s = 0
    r = encoder.N - encoder.F

    # Read initial F bytes
    data_pos = 0
    length = 0
    while length < encoder.F and data_pos < len(data):
        encoder.text_buf[r + length] = data[data_pos]
        data_pos += 1
        length += 1

    if length == 0:
        return None

    # Insert initial strings
    for i in range(1, encoder.F + 1):
        encoder._insert_node(r - i)

    encoder._insert_node(r)

    while length > 0:
        if encoder.match_length > length:
            encoder.match_length = length

        if encoder.match_length <= encoder.THRESHOLD:
            encoder.match_length = 1
            code_buf[0] |= mask
            code_buf[code_buf_ptr] = encoder.text_buf[r]
            code_buf_ptr += 1
        else:
            code_buf[code_buf_ptr] = encoder.match_position & 0xFF
            code_buf_ptr += 1
            code_buf[code_buf_ptr] = (
                ((encoder.match_position >> 4) & 0xF0) |
                (encoder.match_length - (encoder.THRESHOLD + 1))
            )
            code_buf_ptr += 1

        mask <<= 1

        if mask == 0x100:
            compressed.extend(code_buf[:code_buf_ptr])
            code_buf[0] = 0
            code_buf_ptr = 1
            mask = 1

        last_match_length = encoder.match_length
        i = 0

        while i < last_match_length and data_pos < len(data):
            encoder._delete_node(s)
            c = data[data_pos]
            data_pos += 1
            encoder.text_buf[s] = c

            if s < encoder.F - 1:
                encoder.text_buf[s + encoder.N] = c

            s = (s + 1) & (encoder.N - 1)
            r = (r + 1) & (encoder.N - 1)

            encoder._insert_node(r)
            i += 1

        while i < last_match_length:
            encoder._delete_node(s)
            s = (s + 1) & (encoder.N - 1)
            r = (r + 1) & (encoder.N - 1)
            length -= 1
            if length:
                encoder._insert_node(r)
            i += 1

    if code_buf_ptr > 1:
        compressed.extend(code_buf[:code_buf_ptr])

    return bytes(compressed)

def decompress(compressed_data: bytes, settings: Optional[LzssSettings] = None) -> Optional[bytes]:
    """
    Decompress LZSS compressed data.

    Args:
        compressed_data: LZSS compressed bytes
        settings: LzssSettings instance with decompression parameters

    Returns:
        Decompressed data or None if input data is empty
    """
    settings = settings if settings else LzssSettings()
    
    if not compressed_data:
        return None

    text_buf = array.array('B', [settings.frame_fill] * (settings.frame_size + settings.max_match_length - 1))
    decompressed = bytearray()

    N = settings.frame_size
    F = settings.max_match_length
    THRESHOLD = settings.min_match_length

    r = N - F
    flags = 0
    data_pos = 0

    while data_pos < len(compressed_data):
        flags >>= 1
        if (flags & 0x100) == 0:
            if data_pos < len(compressed_data):
                flags = compressed_data[data_pos] | 0xFF00
                data_pos += 1
            else:
                break

        if flags & 1:
            if data_pos < len(compressed_data):
                c = compressed_data[data_pos]
                data_pos += 1
                decompressed.append(c)
                text_buf[r] = c
                r = (r + 1) & (N - 1)
            else:
                break
        else:
            if data_pos + 1 < len(compressed_data):
                i = compressed_data[data_pos]
                j = compressed_data[data_pos + 1]
                data_pos += 2

                i |= ((j & 0xF0) << 4)
                j = (j & 0x0F) + THRESHOLD

                for k in range(j + 1):
                    c = text_buf[(i + k) & (N - 1)]
                    decompressed.append(c)
                    text_buf[r] = c
                    r = (r + 1) & (N - 1)
            else:
                break

    return bytes(decompressed)

File: LzssCompression.go
package lzss

// LzssSettings and DefaultSettings implementations remain unchanged
type LzssSettings struct {
	FrameSize      int
	FrameFill      byte
	FrameInitPos   int
	MaxMatchLength int
	MinMatchLength int
}

func DefaultSettings() *LzssSettings {
	return &LzssSettings{
		FrameSize:      0x1000,
		FrameFill:      0,
		FrameInitPos:   0xFEE,
		MaxMatchLength: 0x12,
		MinMatchLength: 2,
	}
}

// LZSSEncoder implements LZSS compression
type LZSSEncoder struct {
	settings      *LzssSettings
	N             int    // Size of ring buffer
	F             int    // Upper limit for match_length
	THRESHOLD     int    // Minimum length threshold for encoding
	NIL           int    // Index for root of binary search trees
	PADDING       byte   // Padding character
	textBuf       []byte // Ring buffer
	lchild        []int  // Left children
	rchild        []int  // Right children
	parent        []int  // Parent nodes
	matchPosition int    // Current match position
	matchLength   int    // Current match length
}

// NewEncoder creates a new LZSS encoder with given settings
func NewEncoder(settings *LzssSettings) *LZSSEncoder {
	if settings == nil {
		settings = DefaultSettings()
	}

	e := &LZSSEncoder{
		settings:  settings,
		N:         settings.FrameSize,
		F:         settings.MaxMatchLength,
		THRESHOLD: settings.MinMatchLength,
		NIL:       settings.FrameSize,
		PADDING:   settings.FrameFill,
	}

	// Initialize buffers
	e.textBuf = make([]byte, e.N+e.F-1)
	for i := range e.textBuf {
		e.textBuf[i] = e.PADDING
	}

	e.lchild = make([]int, e.N+1)
	e.rchild = make([]int, e.N+257)
	e.parent = make([]int, e.N+1)

	e.initState()
	return e
}

func (e *LZSSEncoder) initState() {
	// Initialize encoding state and trees
	for i := e.N + 1; i < e.N+257; i++ {
		e.rchild[i] = e.NIL
	}
	for i := 0; i < e.N; i++ {
		e.parent[i] = e.NIL
	}
}

func (e *LZSSEncoder) insertNode(r int) {
	var i int
	var p int
	var cmp int

	key := e.textBuf[r : r+e.F]
	p = e.N + 1 + int(key[0])
	e.rchild[r] = e.NIL
	e.lchild[r] = e.NIL
	e.matchLength = 0

	for {
		if cmp >= 0 {
			if e.rchild[p] != e.NIL {
				p = e.rchild[p]
			} else {
				e.rchild[p] = r
				e.parent[r] = p
				return
			}
		} else {
			if e.lchild[p] != e.NIL {
				p = e.lchild[p]
			} else {
				e.lchild[p] = r
				e.parent[r] = p
				return
			}
		}

		i = 1
		for i < e.F {
			cmp = int(key[i]) - int(e.textBuf[p+i])
			if cmp != 0 {
				break
			}
			i++
		}

		if i > e.matchLength {
			e.matchPosition = p
			e.matchLength = i
			if i >= e.F {
				break
			}
		}
	}

	e.parent[r] = e.parent[p]
	e.lchild[r] = e.lchild[p]
	e.rchild[r] = e.rchild[p]
	e.parent[e.lchild[p]] = r
	e.parent[e.rchild[p]] = r

	if e.rchild[e.parent[p]] == p {
		e.rchild[e.parent[p]] = r
	} else {
		e.lchild[e.parent[p]] = r
	}
	e.parent[p] = e.NIL
}

func (e *LZSSEncoder) deleteNode(p int) {
	var q int

	if e.parent[p] == e.NIL {
		return
	}

	if e.rchild[p] == e.NIL {
		q = e.lchild[p]
	} else if e.lchild[p] == e.NIL {
		q = e.rchild[p]
	} else {
		q = e.lchild[p]
		if e.rchild[q] != e.NIL {
			for e.rchild[q] != e.NIL {
				q = e.rchild[q]
			}
			e.rchild[e.parent[q]] = e.lchild[q]
			e.parent[e.lchild[q]] = e.parent[q]
			e.lchild[q] = e.lchild[p]
			e.parent[e.lchild[p]] = q
		}
		e.rchild[q] = e.rchild[p]
		e.parent[e.rchild[p]] = q
	}

	e.parent[q] = e.parent[p]
	if e.rchild[e.parent[p]] == p {
		e.rchild[e.parent[p]] = q
	} else {
		e.lchild[e.parent[p]] = q
	}
	e.parent[p] = e.NIL
}

// Compress compresses input data using LZSS algorithm
func Compress(data []byte, settings *LzssSettings) []byte {
	if len(data) == 0 {
		return nil
	}

	if settings == nil {
		settings = DefaultSettings()
	}

	encoder := NewEncoder(settings)
	codeBuf := make([]byte, 17)
	compressed := make([]byte, 0)

	codeBuf[0] = 0
	codeBufPtr := 1
	mask := uint16(1)

	s := 0
	r := encoder.N - encoder.F

	// Read initial F bytes
	dataPos := 0
	length := 0
	for length < encoder.F && dataPos < len(data) {
		encoder.textBuf[r+length] = data[dataPos]
		dataPos++
		length++
	}

	if length == 0 {
		return nil
	}

	// Insert initial strings
	for i := 1; i <= encoder.F; i++ {
		encoder.insertNode(r - i)
	}
	encoder.insertNode(r)

	for length > 0 {
		if encoder.matchLength > length {
			encoder.matchLength = length
		}

		if encoder.matchLength <= encoder.THRESHOLD {
			encoder.matchLength = 1
			codeBuf[0] |= byte(mask)
			codeBuf[codeBufPtr] = encoder.textBuf[r]
			codeBufPtr++
		} else {
			codeBuf[codeBufPtr] = byte(encoder.matchPosition & 0xFF)
			codeBufPtr++
			codeBuf[codeBufPtr] = byte(((encoder.matchPosition >> 4) & 0xF0) |
				(encoder.matchLength - (encoder.THRESHOLD + 1)))
			codeBufPtr++
		}

		mask <<= 1

		if mask == 0x100 {
			compressed = append(compressed, codeBuf[:codeBufPtr]...)
			codeBuf[0] = 0
			codeBufPtr = 1
			mask = 1
		}

		lastMatchLength := encoder.matchLength
		i := 0

		for i < lastMatchLength && dataPos < len(data) {
			encoder.deleteNode(s)
			c := data[dataPos]
			dataPos++
			encoder.textBuf[s] = c

			if s < encoder.F-1 {
				encoder.textBuf[s+encoder.N] = c
			}

			s = (s + 1) & (encoder.N - 1)
			r = (r + 1) & (encoder.N - 1)

			encoder.insertNode(r)
			i++
		}

		for i < lastMatchLength {
			encoder.deleteNode(s)
			s = (s + 1) & (encoder.N - 1)
			r = (r + 1) & (encoder.N - 1)
			length--
			if length > 0 {
				encoder.insertNode(r)
			}
			i++
		}
	}

	if codeBufPtr > 1 {
		compressed = append(compressed, codeBuf[:codeBufPtr]...)
	}

	return compressed
}

// Decompress decompresses LZSS compressed data
func Decompress(compressedData []byte, settings *LzssSettings) []byte {
	if len(compressedData) == 0 {
		return nil
	}

	if settings == nil {
		settings = DefaultSettings()
	}

	textBuf := make([]byte, settings.FrameSize+settings.MaxMatchLength-1)
	for i := range textBuf {
		textBuf[i] = settings.FrameFill
	}

	decompressed := make([]byte, 0)

	N := settings.FrameSize
	F := settings.MaxMatchLength
	THRESHOLD := settings.MinMatchLength

	r := N - F
	flags := uint16(0)
	dataPos := 0

	for dataPos < len(compressedData) {
		flags >>= 1
		if (flags & 0x100) == 0 {
			if dataPos < len(compressedData) {
				flags = uint16(compressedData[dataPos]) | 0xFF00
				dataPos++
			} else {
				break
			}
		}

		if flags&1 != 0 {
			if dataPos < len(compressedData) {
				c := compressedData[dataPos]
				dataPos++
				decompressed = append(decompressed, c)
				textBuf[r] = c
				r = (r + 1) & (N - 1)
			} else {
				break
			}
		} else {
			if dataPos+1 < len(compressedData) {
				i := int(compressedData[dataPos])
				j := int(compressedData[dataPos+1])
				dataPos += 2

				i |= ((j & 0xF0) << 4)
				j = (j & 0x0F) + THRESHOLD

				for k := 0; k <= j; k++ {
					c := textBuf[(i+k)&(N-1)]
					decompressed = append(decompressed, c)
					textBuf[r] = c
					r = (r + 1) & (N - 1)
				}
			} else {
				break
			}
		}
	}

	return decompressed
}
